### PR TITLE
Fixes #4468 and #4476

### LINKED ISSUE #4468: Support decoding atom queries as SMARTS in data SGroups
Using an `MRV SMA` atom property is a nice solution for capturing complex query information in a V2000 mol block: https://github.com/rdkit/rdkit/blob/69b143edd059612c2f28ad898fa2d87dc1525e6f/Code/GraphMol/FileParsers/MolFileParser.cpp#L809

This is not available for V3000 (which doesn't have atom properties), but we could provide similar functionality using data SGroups.
Here's a suggested format:
```
query
  Mrv2108 07152116012D          

  0  0  0     0  0            999 V3000
M  V30 BEGIN CTAB
M  V30 COUNTS 2 1 1 0 0
M  V30 BEGIN ATOM
M  V30 1 C -0.8333 4.5421 0 0
M  V30 2 C 0.5003 5.3121 0 0
M  V30 END ATOM
M  V30 BEGIN BOND
M  V30 1 1 1 2
M  V30 END BOND
M  V30 BEGIN SGROUP
M  V30 1 DAT 0 ATOMS=(1 2) -
M  V30 FIELDDISP="    0.0000    0.0000    DR    ALL  0       0" -
M  V30 QUERYTYPE=SMARTSQ QUERYOP== FIELDDATA=[#6;R]
M  V30 END SGROUP
M  V30 END CTAB
M  END
```

One nice thing about this approach is that it will work with both V3000 and V2000 mol blocks.

### PR DESCRIPTION
**Pull request by greglandrum**
Changes here:
- #4468: support reading SMARTS queries on atoms from SGroups in V2000 and V3000 mol blocks. These are indicated using   `QUERYTYPE=SMARTSQ` or `QUERYTYPE=SQ`.
- #4476: read Data SGroup (i.e. SDT) properties from V2000 mol blocks even if the FIELDNAME is empty
- Changes the behavior of `RWMol.replaceAtom()` to not remove SubstanceGroups involving the changed atom. This isn't necessary and doesn't really make sense.

### CODE REVIEW

**Comment by greglandrum** [comment]
@ricrogz : this isn't quite ready to merge yet, but I'd appreciate your feedback on the approach.
This is connected to SHARED-8091

**Comment by ricrogz** [comment]
Yeah, I think this makes sense 👍

**Review by ricrogz — APPROVED** [review]
Just a couple of notes. Looks good otherwise.

File: Code/GraphMol/FileParsers/MolFileParser.cpp [hunk 1]
```diff
@@ -2724,6 +2724,65 @@ void processMrvCoordinateBond(RWMol &mol, const SubstanceGroup &sg) {
   }
 }
 
+void processSMARTSQ(RWMol &mol, const SubstanceGroup &sg) {
+  std::string field;
+  if (sg.getPropIfPresent("QUERYOP", field) && field != "=") {
+    BOOST_LOG(rdWarningLog) << "unrecognized QUERYOP '" << field
+                            << "' for SMARTSQ. Query ignored." << std::endl;
+    return;
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/FileParsers/MolFileParser.cpp:2732]: > This makes "QUERYOP" optional, is that intentional? Or should this be:

Yeah, I figure we can assume that it's `=` if it's missing.

**Pull request merged**